### PR TITLE
Binary targets proposal implementation

### PR DESCRIPTION
**Pull request by hartbit**

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -109,14 +109,14 @@ let package = Package(
             /** Abstractions for common operations, should migrate to TSCBasic */
             name: "TSCUtility",
             dependencies: ["TSCBasic"]),
-        
+
```
Comment (neonichu) [inline Package.swift]: 😱

File: Sources/Commands/SwiftPackageTool.swift [hunk 2]
```diff
@@ -48,12 +48,12 @@ public class SwiftPackageTool: SwiftTool<PackageToolOptions> {
 
             switch configMode {
             case .getMirror:
-                guard let packageURL = options.configOptions.packageURL else {
-                    diagnostics.emit(.missingRequiredArg("--package-url"))
+                guard let originalURL = options.configOptions.originalURL else {
```
Comment (neonichu) [inline Sources/Commands/SwiftPackageTool.swift]: Should we split the mirror changes out into its own PR?

File: Sources/PackageLoading/ManifestLoader.swift [hunk 3]
```diff
@@ -12,7 +12,7 @@ import TSCBasic
 import PackageModel
 import TSCUtility
 import SPMLLBuild
-import class Foundation.ProcessInfo
```
Comment (neonichu) [inline Sources/PackageLoading/ManifestLoader.swift]: Did this need to change?

File: Sources/PackageLoading/ManifestLoader.swift [hunk 4]
```diff
@@ -355,6 +374,32 @@ public final class ManifestLoader: ManifestLoaderProtocol {
...
+            let isRemote = target.url != nil
+            let validSchemes = ["https"]
+            guard !isRemote || (location.scheme.map({ validSchemes.contains($0) }) ?? false) else {
+                throw ManifestParseError.invalidBinaryURLScheme(
+                    targetName: target.name,
+                    validSchemes: validSchemes
+                )
+            }
+
+            let validExtensions = isRemote ? ["zip"] : ["zip", "xcframework"]
```
Comment (neonichu) [inline Sources/PackageLoading/ManifestLoader.swift]: For local ones, we did want to disallow zip archives, right?

File: Sources/TSCUtility/Git.swift [hunk 5]
```diff
@@ -69,6 +69,16 @@ public class Git {
         }
     }
 
+    /// Returns the SHA-1 hash of a file.
+    public static func hashObject(at path: String) throws -> String {
```
Comment (neonichu) [inline Sources/TSCUtility/Git.swift]: Is this still being used?

**Comment by mackoj** [comment]
Should it be limited to zip file type ? I open a PR with support for tar based files(#2481) in hopes we could use tar based or zip for binary target.
Cheers,
